# BGW-xWAS-SS: Bayesian Genome-wide xWAS Using xQTL Summary Statistics

**BGW-xWAS-SS** is developed for leveraging **summary-level** **cis-** and **trans-** xQTL data based on a Bayesian variable selection model to predict genetically regulated molecular quantitative traits and conduct follow-up xWAS analysis. The product of the estimated **cis-** and **trans-** xQTL effect sizes and their corresponding causal posterior probabilities (CPP) of being an xQTL will be taken as variant weights (i.e., **xQTL weights**) to conduct **xWAS** –– gene-based association test with test GWAS data. This tool is implemented through several command-line scripts described in this manual. Example shell commands are provided in `./Run_BGW.sh`.

Please cite our BGW-TWAS papers if you use the tool:
>[*Bayesian Genome-wide TWAS Method to Leverage both cis- and trans-eQTL Information through Summary Statistics.* 2020 AJHG.](https://www.cell.com/ajhg/pdfExtended/S0002-9297(20)30291-3)

Please contact **Jingjing Yang (<jingjing.yang@emory.edu>)** if there is any issue.


---

- [Software Installation](#software-installation)
- [Input Files](#input-files)
	- [1. Individual-level Training Data Files](#1.-individual-level-training-data-files)
	- [2. xQTL Summary Statistic Files](#2.-xQTL-summary-statistic-files)
	- [3. Individual-level Test GWAS Data Files](#3.-individual-level-test-gwas-data-files)
- [Example Usage](#example-usage)
	- [1. Set up Tool Directories and Input Arguments](#1.-set-up-tool-directories-and-input-arguments)
	- [2. Generate xQTL Summary Statistics](#2.-generate-xQTL-summary-statistics)
	- [3. Prune Genome Blocks](#3.-prune-genome-blocks)
	- [4. Training xQTL Weights by EM-MCMC Algorithm](#4.-training-xQTL-weights-by-em-mcmc-algorithm)
	- [5. Predict Genetically Regulated Molecular Traits](#5.-predict-genetically-regulated-molecular-traits)
	- [6. Gene-based Association Test](#6.-gene-based-association-test)

---

## Software Installation

### 1. Compile **Estep_mcmc**
* Install required C++ libraries **zlib**, **gsl**, **eigen3**, **lapack**, **atlas**, **blas** that are used to develop this tool. Please install these libraries to your system and include the library path `-I[path to libraries]` accordingly in the C++ compilation command line in the `./Makefile`.

* Compile C++ library `./libStatGen/libStatGen.a` under your system by using the following commands:

```
cd BGW-TWAS-SS/libStatGen/;
make clean;
make
```

* Compile C++ source code for the executible file *./bin/Estep_mcmc* that will be used to run the Estep MCMC algorithm to estimate eQTL effect sizes and the posterior causal probabilities (PCP) to be an eQTL, by using the following commands under `BGW-TWAS/` directory:

```
cd BGW-TWAS-SS/;
make clean ;
make
```

* Even though an executible file *BGW-TWAS-SS/bin/Estep_mcmc* compiled on our computing cluster is provided on GITHUB, please still compile one for your own system.

### 2. Additional Requirements
* Linux tool: [**TABIX**](https://www.htslib.org/doc/tabix.html)

## Input Files

Input files of **BGW-TWAS-SS** tool are all tab-seperated text files:

* **Individual-level training transcriptomic data files** include gene expression file, genotype files of training samples, and a list of filenames for training genotype files, which are required for training Bayesian gene expression prediction models to estimate **cis-** and **trans-** eQTL effect sizes and their corresponding posterior causal probabilities of being an eQTL.
* **Summary-level training eQTL data files** include Z-score statistic files and reference LD files. If individual-level reference transcriptomic data are available, such summary-level eQTL data files will be generated by the **BGW-TWAS-SS** tool in the first step.
* **Individual-level GWAS data files** of test samples include genotype VCF files and a phenotype file.
* **Summary-level GWAS data files** include a text file of Z-score statistics by single variant GWAS test.


### 1. Individual-level Training Data Files
#### 1.1. Gene Annotation and Molecular Quantitative Trait File

* Provide data of the molecular quantitative traits of training samples as in `./Example/ExampleData/Gene_Exp_example.txt`, with one gene per row, and one sample per column starting from the 6th column. The first five columns are required to be gene annotation information including chromosome number, starting position, ending position, Gene ID, and Gene Name or second set of Gene ID. 
* Gene annotations should be for the corresponding protein coding gene, or nearest gene.
* The molecular quantitative traits in this file should be the residuals of a linear regression model that regresses out other confounding variables such as age, sex, top 5 genotype PCs from the `log2(TPM)` of raw sequencing data.
* Sample IDs should be consistent as those in the training VCF genotype files.
* Example row of the first 6 columns of gene `ABCA7` is as follows:

| CHROM |	GeneStart |	GeneEnd |	    TargetID     | GeneName |	 ROS20275399 |
| ----- | --------- | ------- | ---------------- | -------- | ------------ |
|  19	  |  1040101	| 1065571 |	 ENSG00000064687 |	 ABCA7  | 0.6707739044 |


#### 1.2. VCF Genotype Files

* **[VCF Genotype files](http://samtools.github.io/hts-specs/VCFv4.1.pdf)** are used for training prediction models for molecular quantitative traits. The VCF genotype files should be one per genome block (variants of the same chromosome should be in the same block), sorted by position, and then zipped by `bgzip`. File names are of `[filehead].vcf.gz` format, e.g., `./Example/ExampleData/genotype_data_files/Cis_geno_block.vcf.gz`. 
	- All VCF genotype files should be put under the same directory, `${geno_dir}`, which should also be provided. For example, see genotype files `*_geno_block.vcf.gz` under the directory `./Example/ExampleData/genotype_data_files/`.
	* **BGW-xWAS-SS** tool will determine **cis-** or **trans-** SNPs based on the gene start/end information provided in the gene expression file. No need to manually put cis- and trans- SNPs in different genome blocks.
	* Genotype files are supposed to be segmented based on LD per ancestry. Genome blocks are expected to be approximately independent with ~5K-10K SNPs per block. 
	*  Example genome block information for ASN, AFR, and EUR populations generated by **[LDetect](https://bitbucket.org/nygcresearch/ldetect/src/master/)** are provided under `/Example/ExampleData/lddetect_block_bed/` with base pair positions in hg19 format.

* **List of file heads** of VCF genotype files as in `./Example/ExampleData/geno_block_filehead.txt` is required. Each row is the file head of one VCF file of one genome block as in `[filehead].vcf.gz`. Note that the VCF file extension suffix `.vcf.gz` should not be included.

*******************************************************


### 2. xQTL Summary Statistic Files

#### 2.1. xQTL Zscore Files 
* **Zscore Summary Statistic** files generated by single variant tests of the association between each genetic variant and each molecular quantitative trait (i.e., per gene), one file per genome block, bgzipped, with names in the format of `[filehead].Zscore.txt.gz`. See example files under `./Example/ExampleWorkDir/ABCA7_Zscores/`.
	* Each row in the Zscore file represents the annotation and summary statistic information per genetic variant.
	* A total of 10 columns seperated by tabs are contained in the Zscore file, including chromorome number (\#CHROM), basepair position (POS), genetic variant ID (ID), reference allele (REF), alternative allele (ALT), training sample size (N), minor allele frequency (MAF), Z-score statistic value (Z_SCORE), marginal effect size (mBeta), and p-value for single variant test of the association between the target genetic variant and the molecular quantitative trait (PVALUE).
	* **A set of Zscore summary statistic files are needed for each gene.**
	* Example Zscore file header and first 2 SNPs:

| #CHROM |	POS |	ID |	    REF    | ALT | N | MAF |	 Z_SCORE | mBeta | PVALUE
| ----- | --------- | ------- | ---------------- | -------- | ------------ | ---------------- | -------- | ------------ |-------- | ------------ |
|  15	| 38530777 |	 rs1522784;15:38530777 | C | T | 499 | 0.031 | 	 -1.6e+00 | -7.5e-02  |    9.1e-02 |
| 15	| 38531521 |	 rs8041343;15:38531521 | C | T | 499 | 0.107 | 	 1.1e-01 | 5.1e-03     |  9.0e-01 |

####  2.2. LD Files 
* **Genotype correlation (LD)** files, one per genome block, bgzipped, with names in the format of `[filehead].LDcorr.txt.gz`. See example files under `./Example/ExampleData/LDdir/`.

	* Each row in the LD file represents the information of one target genetic variant (orderred by their base pair positions), and correlations between this target genetic variant and next xx number of genetic variants with positions within a specified window (e.g., POS + 1MB). 

	* A total of 9 columns seperated by tabs are contained in the LD file, including row number (\#ORDER), chromosome number (CHROM), base pair position (POS), genetic variant ID (ID), reference allele (REF), alternative allele (ALT), training sample size (N), minor allele frequency (MAF), correlations with genetic variants within base pair positions within a specified window beyong this genetic variant (CORR, correlations are seperated by `,`, starting with correlation with itself `1`).
	* **Only one set of LD files are needed for all genes.**  
	* Example LD file header and first 2 SNPs, with `-7.1e-02` denoting the correlation between these two example SNPs:

| #ORDER | CHROM |	POS |	ID |	    REF    | ALT | N | MAF |	 CORR |
| ----- | --------- | ------- | ---------------- | -------- | ------------ | ---------------- | -------- | ------------ |
| 0	  |  15	| 38530777 |	 rs1522784;15:38530777 | C | T | 499 | 0.031 | 	 1.0e+00,-7.1e-02,-5.6e-02,... |
| 1	  | 15	| 38531521 |	 rs8041343;15:38531521 |C | T | 499 | 0.107 | 	 1.0e+00,9.5e-01,9.5e-01,... |

#### 2.3. Gene Annotation File

* First 5 columns of the molecular trait file as in `./Example/ExampleData/Gene_Exp_example.txt`, with one gene per row. 
	* Gene annotations should be for the corresponding protein coding gene, or nearest gene. Information about the chromosome number (CHROM), gene transcription starting site (GeneStart) and gene trascription termination site (GeneEnd), gene ID such as ensemble ID (TargetID), and gene name (GeneName) are provided.
	* Example row of gene `ABCA7` is as follows:

| CHROM |	GeneStart |	GeneEnd |	    TargetID     | GeneName |
| ----- | --------- | ------- | ---------------- | -------- | 
|  19	  |  1040101	| 1065571 |	 ENSG00000064687 |	 ABCA7  |


#### 2.4. Remarks
* Either individual-level moledular quantitative trait and genotype files or xQTL summary statistic files are needed for training xQTL effect sizes in the Bayesian variable selection model.
* If only individual-level training data are available, **BGW-xWAS-SS** tool will first generate xQTL summary statistic files and use these files to estimate xQTL weights for xWAS.

*******************************************************


### 3. Individual-level Test GWAS Data Files

* **Test [VCF Genotype files](http://samtools.github.io/hts-specs/VCFv4.1.pdf)** are required for predicting genetically regulated molecular traits of test samples. Different from the training VCF genotype files, these test genotype files should be one per chromosome containing `_CHR[chr_num]_` in their file name, sorted by position, zipped by `bgzip`, and indexed by `TABIX` tool. For example, see `./Example/ExampleData/genotype_data_files/Test_CHR*_geno.vcf.gz`.

* **Test phenotype file** contains two columns without header: the first column is a list of sample IDs that have genotype data in the test VCF files, and the second column is phenotype values. Sample IDs in the phenotype file should be consistent as those in the test VCF files. Second column will not be used for predicting genetically regulated molecular trait values, thus psuedo values such as all `1s` can be used. For example, see `./Example/ExampleData/Test_pheno.txt`


## Example Usage

### 1. Set up Tool Directories and Input Arguments

```
## Variables for training xQTL weights.

### Please update tool directory and use different working and LD file directories from below
BGW_dir=~/GIT/BGW-xWAS-SS # Tool directory
wkdir=${BGW_dir}/Example/ExampleWorkDir
Genome_Seg_Filehead=${BGW_dir}/Example/ExampleData/geno_block_filehead.txt
gene_name=ABCA7 # target gene
GeneInfo=${BGW_dir}/Example/ExampleData/Gene_Exp_example.txt # Gene annotation file or Molecular trait file


### Directories for using summary-level xQTL data
ZScore_dir=${wkdir}/${gene_name}_Zscores # Zscore statistic files
LDdir=${BGW_dir}/Example/ExampleData/LDdir # Shared by all genes

num_cores=2 # Number of cores to be used

### Directories and variables needed for using individual-level genotype data
geno_dir=${BGW_dir}/Example/ExampleData/genotype_data_files # Genotype VCF directory
GTfield=GT # specify genotype field "DS" or the corresponding field name in the VCF file for dosage genotype

## Varables for selecting genome blocks 
p_thresh=0.001 # p-value threshold
max_blocks=50 # maximum blocks

## Variables for running EM-MCMC with selected genome blocks
N=499 # Training sample size
hfile=${BGW_dir}/Example/hypval.txt
PCP_thresh=0.0001

## Variables for predicting genetically regulated molecular trait using individual-level test genotype data
BGW_weight=${wkdir}/${gene_name}_BGW_eQTL_weights.txt
test_geno_dir=${BGW_dir}/Example/ExampleData/genotype_data_files
test_geno_filehead=${BGW_dir}/Example/ExampleData/test_geno_filehead.txt
test_pheno=${BGW_dir}/Example/ExampleData/Test_pheno.txt
GTfield_test=GT # or DS if doseage data to be used
```

#### Remarks
* It is important to include the complete directory starting with `/home/` for all files.
* Please set up working directory `${wkdir}` and LD file directory `${LDdir}` differently from the above examples, so that you may examine how your test outputs  be different from the provided example outputs.
* LD file directory `${LDdir}` stays the same for all genes.
* Please do not directly run the `./Run_BGW.sh` script. Instead, test the bash commands of the following steps one by one.

### 2. Generate xQTL Summary Statistics 
- Generate xQTL summary statistics if only individual-level training data are provided

#### 2.1. Shell script `get_sumstat.sh` will generate single variant xQTL summary statistics (i.e., Score Statistics) in required formats.

#### 2.2. Input arguments
- `--BGW_dir` : Specify the directory of BGW-TWAS tool
- `--wkdir` : Specify a working directory with writing access
- `--LDdir` : Directory of all LD files
- `--Genome_Seg_Filehead` : Genome segmentation file
- `-GeneInfo` : Gene annotation or molecular trait file directory
- `--gene_name` : Gene name as in the 5th column of `-GeneExpFile`
- `--geno_dir` : Directory of individual-level genotype files
- `--GTfield` : Specify the genotype format in the training vcf file that should be used. Default `GT` for assayed genotype. Alternative value `DS` for the inputed dosage genotype field by [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html#!), or other genotype _FIELD_ name as used in the VCF file.
- `--num_cores` : Number of parallele sessions, default `1`.
- `--clean_output` : Whether to delete all intermediate outcomes, taking the input value of `1` for deleting or `0` keeping all intermediate files for testing purpose.

#### 2.3. Example command:
```
${BGW_dir}/bin/get_sumstat.sh --BGW_dir ${BGW_dir} \
--wkdir ${wkdir} --gene_name ${gene_name} --GeneInfo ${GeneInfo} \
--geno_dir ${geno_dir} --LDdir ${LDdir} --Genome_Seg_Filehead ${Genome_Seg_Filehead} \
--GTfield ${GTfield} --num_cores ${num_cores} --clean_output 1
```

#### 2.4. Output files
* **xQTL Zscore statistic files** are generated under `${wkdir}/${gene_name}_Zscores/`, one file per genome block.

* **LD files** will be generated for all genome blocks, one file per genome block, if LD files under `${LDdir}/` are not provided. The same set of LD files will only be generated once and used for all genes.


#### Remarks
* Rare variants with MAF <1% will be excluded from analysis by default.
* These Zscore statistics and LD files will be used for implementing the *EM-MCMC* algorithm to estimate xQTL weights based on the Bayesian variable selection model.

### 3.  Prune Genome Blocks
#### 3.1. Script `prune.sh` will select a subset of genome blocks (up to `${max_blocks}`) to run the EM-MCMC algorithm, 
* Cis blocks are always selected. 
* Trans blocks with minimum single-variant xQTL `p-value < ${p_thresh}` will first be ranked by the smallest p-value within block (from the smallest to the largest), and then top ranked trans blocks will be selected up to `${max_blocks}`.


#### 3.2. Input arguments
- `--wkdir` : Specify a working directory
- `--gene_name` : Gene name that should be the same used in `GeneExpFile`
- `-GeneExpFile` : Gene expression file directory
- `--Genome_Seg_Filehead` : Directory of the file containing a list of fileheads of segmented genotype files
- `--ZScore_dir` : Specify summary score statistic file directory, default `${wkdir}/${gene_name}_scores`
- `--p_thresh` : Specify p-value threshold for pruning, default `1e-5`
- `--max_blocks` : Specify the maximum number of genome blocks for jointly training gene expression prediction models, default `50`

#### 3.3. Example command:
```
${BGW_dir}/bin/prune.sh --wkdir ${wkdir} --gene_name ${gene_name} \
--GeneInfo ${GeneInfo} --Genome_Seg_Filehead ${Genome_Seg_Filehead} \
--ZScore_dir ${ZScore_dir} \
--p_thresh ${p_thresh} --max_blocks ${max_blocks} --clean_output 1
```


#### 3.4. Output files
* A list of filehead of selected genome blocks and the corresponding minimum within-block single xQTL analysis p-values are given (one genome block per row) in the file of `${wkdir}/${gene_name}_select_filehead.txt`, which will be used to train xQTL weights by **EM-MCMC** algorithm.

### 4. Training xQTL Weights by EM-MCMC Algorithm

#### 4.1. Script `EM-MCMC.sh` will train xQTL weights
* Summary Zscore statistics and LD files for selected genome blocks will be used to run EM-MCMC algorithm, 
* xQTL weights will be estimated based on the Bayesian variable selection model. 
* Weights of xQTL with Bayesian causal posterior probability (CPP) `CPP>${CPP_thresh}` will be saved for the follow-up xWAS analysis.

#### 4.2. Input arguments
- `--BGW_dir` : Directory of BGW-TWAS tool
- `--wkdir` : Working directory with writting access
- `-GeneExpFile` : Gene expression file directory
- `--gene_name` : Study gene name that should be the same used in `GeneExpFile`
- `--LDdir` : Directory of all LD files with writting access
- `--Score_dir` : Specify summary score statistic file directory, default `${wkdir}/${gene_name}_scores`
- `--select_filehead` : File with selected genome block fileheads, default `${wkdir}/${gene_name}_select_filehead.txt`
- `--N` : Number of sample size used in Step 1 to generate eQTL summary statistics
- `--hfile` : Hyper parameter file as in `./Example/hypval.txt`, specifing the prior causal probability (_pi_) and effect size variance (_sigma2_) for cis (row 1) and trans (row 2) eQTL
- `--em` : Number of EM iterations. Default `3`.
- `--burnin` : Number of burnin MCMC iterations. Default `10000`.
- `--Nmcmc` : Number of MCMC iterations. Default `10000`.
- `--PCP_thresh` : PCP threshold for selecting eQTL with `PCP > ${PCP_thresh}` for predicting GReX and association study. Default `0.0001`.
- `--num_cores` : Specify the number of parallele sessions, default `1`.
- `--clean_output` : Whether to delete all intermediate outcomes, taking the input value of `1` for deleting or `0` keeping all intermediate files for testing purpose.

#### 4.3. Example commands:
```
select_filehead=${wkdir}/${gene_name}_select_filehead.txt

${BGW_dir}/bin/EM-MCMC.sh  --BGW_dir ${BGW_dir} \
--wkdir ${wkdir} --gene_name ${gene_name} \
--GeneExpFile ${GeneExpFile} --LDdir ${LDdir} \
--Score_dir ${Score_dir} --select_filehead ${select_filehead} \
--N ${N} --hfile ${hfile} \
--em 3 --burnin 10000 --Nmcmc 10000 \
--PCP_thresh ${PCP_thresh} --num_cores ${num_cores} \
--clean_output 0
```


#### 4.4. Output files
* Bayesian estimates of eQTL PCP and effect sizes from the final EM-MCMC iteration will saved in the output `${gene_name}_BGW_eQTL_weights.txt` file under specified `${wkdir}` that lists all SNPs with `PCP>${PCP_thresh}`.

* The BGW weight file (`${gene_name}_BGW_eQTL_weights.txt`) will be used for predicting GReX values with individual-level GWAS data as in Step 4 or conducting gene-based association test with GWAS summary statistics. See instructions in Step 5 for TWAS procedure.



#### Remarks
* Intermediate output will be deleted unless with input argument `--clean_output 0`. Keeping intermediate outputs is recommended only for testing purpose.
*  Additional arguments can be tuned in the `EM-MCMC.sh` script, but should be done with caution. These arguments are detailed in [Yang et al. 2017](https://github.com/yjingj/bfGWAS/blob/master/bfGWAS_Manual.pdf).


### 5. Predict Genetically Regulated Molecular Traits
Step 4 will use the BGW weight file (`${gene_name}_BGW_eQTL_weights.txt`) generated from Step 3 and provided individual-level GWAS data to predict GReX values for test samples. Both test genotype VCF files (saved per chromosome, each file name containing the pattern of `_CHR[chr_num]_`) and test phenotype file should be provided.

The product of eQTL PCP and effect size will give an expected eQTL effect size that will be used as SNP weight for estimating the GReX values in Step 4.

#### 5.2.  Input arguments
- `--BGW_dir` : Directory of BGW-TWAS tool
- `--wkdir` : Working directory with writting access
- `--gene_name` : Study gene name that should be the same used in `GeneExpFile`
- `--BGW_weight` : Directory of the BGW eQTL weight file
- `--test_geno_dir` : Directory of all test genotype VCF files
- `--test_geno_filehead` : List of fileheads for test genotype VCF files
- `--test_pheno` : Directory of the test phentoype file
- `--GTfield` : Specify the genotype field in the VCF file to be used. Default `GT` for assayed genotype. Alternative value `DS` for the inputed dosage genotype field by [Michigan Imputation Server](https://imputationserver.sph.umich.edu/index.html#!).
- `--num_cores` : Specify the number of parallele sessions, default `1`.
- `--clean_output` : Whether to delete all intermediate outcomes, taking the input value of `1` for deleting or `0` keeping all intermediate files for testing purpose.


#### 5.3. Example commands:

```
${BGW_dir}/bin/Step4_get_test_grex.sh --BGW_dir ${BGW_dir} \
--wkdir ${wkdir} --gene_name ${gene_name} \
--BGW_weight ${BGW_weight} --test_geno_dir ${test_geno_dir} \
--test_geno_filehead ${test_geno_filehead} \
--GTfield ${GTfield_test} --test_pheno ${test_pheno} \
--num_cores ${num_cores}
```

* Intermediate output will be deleted unless with input argument `--clean_output 0`. Keeping intermediate outputs is recommended only for testing purpose.


#### 5.4. Output
* Predicted GReX values are saved in `${wkdir}/${gene_name}_pred_grex.txt`, which can be used to calculate prediction R2 and test the association between _GReX_ and _Phenotype of Interest_ (i.e., TWAS).

* File `${wkdir}/${gene_name}\_sumPCP` contains the sum of posterior causal probabilities (PCP) of all analyzed SNPs, the sum of cis-SNPs, and the sum of trans-SNPs, which are the expected number of total eQTL, cis-eQTL, and trans-eQTL for the target gene.

### 6. Gene-based Association Test

* If using summary-level GWAS data for TWAS, Step 4 will not need to be implemented. One can use the weight files `${wkdir}/${gene_names}_BGW_eQTL_weights.txt` to select the test SNPs presenting in this weight file. 

* Note that the eQTL weights estimated BGW are using only centered gene expressions and genotype data but not standardized to have standard deviation `1`. For a valid TWAS using GWAS summary statistics, the S-PrediXcan test statistic with genotype covariance matrix in the denominator must be used, as the FUSION test statistic with genotype correlation matrix in the denominator assumes eQTL weights are derived from standardized gene expressions and genotype data.

* With summary-level GWAS Z-score statistics <img src="https://render.githubusercontent.com/render/math?math=Z_i, i=1, \cdots, m"> for single SNP tests, eQTL effect sizes (i.e., weights `w`) estimated by BGW-TWAS method for `m` test SNPs
	*  "S-PrediXcan" TWAS Z score statistic is given by
<img src="https://render.githubusercontent.com/render/math?math=\frac{\sum_{i=1}^m \left(w_i \sigma^2_i Z_i \right) } {\sqrt{\mathbf{w'\Sigma w}}}">, with reference LD **Covariance matrix <img src="https://render.githubusercontent.com/render/math?math=\mathbf{\Sigma}">**, and reference genotype variance (diagonal values of the LD covariance matrix) of SNP `i` <img src="https://render.githubusercontent.com/render/math?math=\sigma^2_i">.
	* "FUSION" TWAS Z score statistic is given by <img src="https://render.githubusercontent.com/render/math?math=\frac{\sum_{i=1}^m \left(w_i Z_i\right) } {\sqrt{\mathbf{w'Vw}}}">, with reference LD **Correlation matrix `V`**.
	* One can use "TABIX" tool to extract reference genotype data of the test SNPs from external reference VCF files, for the purpose of calculating the reference covariance or correlation matrix more easily.

* Or one can use our other [TIGAR](https://github.com/yanglab-emory/TIGAR) tool to obtain TWAS results, where both "S-PrediXcan" and "FUSION" TWAS Z score tests were implemented.



* If using individual-level GWAS data for TWAS, Step 4 will need to be implemented to obtain predicted GReX values for all test samples. Then a simple single variant association test between _GReX_ and _Phenotype of Interest_ will result in the TWAS results.


## Options to save disk storage

* By default setting, temporary directories such as `${wkdir}/${gene_name}_scores/output/`, `${wkdir}/${gene_name}_EM_MCMC/`, `${wkdir}/${gene_name}_GReX/`, and intermediate files `${wkdir}/${gene_name}_exp_trait.txt`, `${wkdir}/${gene_name}_exp_var.txt`, `${wkdir}/ABCA7_grex.geno` would be deleted. If not, you can delete those.

* All score statistics files under `${wkdir}/${gene_name}_scores/` contain all single variant eQTL analysis test summary statistics, which is optional to be either saved for other usage or deleted.

* The weight files `${wkdir}/${gene_names}_BGW_eQTL_weights.txt` are the only file needed for TWAS, which can be gzipped to save storage.

